core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java

```diff
@@ -194,8 +194,16 @@ void failTask(Task task, PayloadType payloadType, String errorMsg) {
         throw new TerminateWorkflowException(errorMsg, Workflow.WorkflowStatus.FAILED, task);
     }
 
-    private void failWorkflow(String errorMsg) {
+    @VisibleForTesting
+    void failWorkflow(Workflow workflow, PayloadType payloadType, String errorMsg) {
         LOGGER.error(errorMsg);
-        throw new TerminateWorkflowException(errorMsg);
+        workflow.setReasonForIncompletion(errorMsg);
+        workflow.setStatus(Workflow.WorkflowStatus.FAILED);
+        if (payloadType == PayloadType.TASK_INPUT) {
```

Comment:
From where the `failWorkflow` is called, the `PayloadType` should always be `WORKFLOW_INPUT` or `WORKFLOW_OUTPUT`.

Reply:
yeah the `payloadType` is passed into the `verifyAndUpload` method: https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java#L87 And we always verify both input and output data: https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/execution/DeciderService.java#L566

Reply:
Yes. `ExternalPayloadStorageUtils.verifyAndUpload()` is never called with a `Workflow` object and `PayloadType` as `TASK_INPUT` or `TASK_OUTPUT`. Workflow's input/output should not be manipulated here since the caller might be already manipulating it.

For instance, if the workflow definition has a `failure_workflow` defined, the workflow output is manipulated and setting it to `null` here would cause a `NullPointerException` there.

Long story short, i don't see a reason to update the `failWorkflow` method to fix this issue.

Reply:
I added the logic to setting workflow output or input to `null` in the `failWorkflow` method because in normal circustances (i.e. when input or output size don't exceed threshold), we always do set the input/output to `null`: https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java#L162 So to be consistent, I think it makes sense to have the same logic added for `failWorkflow`: https://github.com/Netflix/conductor/blob/main/core/src/main/java/com/netflix/conductor/core/utils/ExternalPayloadStorageUtils.java#L135